modules/qualitycontrol/QualityController.ts

```diff
@@ -0,0 +1,366 @@
+      this._maxSize = size;
+      this._data = [];
+    }
+  
+    add(item: ISourceStats): void {
+      if (this._data.length >= this._maxSize) {
+        this._data.shift();
+      }
+      this._data.push(item);
+    }
+  
+    get(index: number): ISourceStats | undefined {
+      if (index < 0 || index >= this._data.length) {
+        throw new Error("Index out of bounds");
+      }
+      return this._data[index];
+    }
+
+    size(): number {
+        return this._data.length;
+    }
+}
+
+/**
+ * QualityController class that is responsible for maintaining optimal video quality experience on the local endpoint
+ * by controlling the codec, encode resolution and receive resolution of the remote video streams. It also makes
+ * adjustments based on the outbound and inbound rtp stream stats reported by the underlying peer connection.
+ */
+export default class QualityController {
+    private _codecController: CodecSelection;
+    private _conference: JitsiConference;
+    private _enableAdaptiveMode: boolean;
+    private _encodeTimeStats: Map<string, FixedSizeArray>;
+    private _limitedByCpuTimeout: number | undefined;
+    private _receiveVideoController: ReceiveVideoController;
+    private _rtc: RTC;
+    private _sendVideoController: SendVideoController;
+
+    /**
+     * 
+     * @param {JitsiConference} conference - The JitsiConference instance.
+     * @param {RTC} rtc - RTC instance.
+     * @param {Object} codecSettings - Codec settings for jvb and p2p connections passed to config.js.
+     * @param {boolean} enableAdaptiveMode - Whether the client needs to make runtime adjustments for better video
+     * quality.
+     */
+    constructor(conference: JitsiConference, rtc: RTC, codecSettings: any, enableAdaptiveMode: boolean) {
+        this._conference = conference;
+        this._rtc = rtc;
+        this._codecController = new CodecSelection(conference, codecSettings);
+        this._enableAdaptiveMode = enableAdaptiveMode;
+
+        this._receiveVideoController = new ReceiveVideoController(conference, rtc);
+        this._sendVideoController = new SendVideoController(conference, conference);
+        this._encodeTimeStats = new Map();
+
+        this._conference.on(
+            JitsiConferenceEvents._MEDIA_SESSION_STARTED,
+            (session: JingleSessionPC) => {
+                this._codecController.selectPreferredCodec(session);
+                this._receiveVideoController.onMediaSessionStarted(session);
+                this._sendVideoController.onMediaSessionStarted(session);
+            });
+        this._conference.on(
+                JitsiConferenceEvents._MEDIA_SESSION_ACTIVE_CHANGED,
+                () => this._sendVideoController.configureConstraintsForLocalSources());
+        this._conference.on(
+            JitsiConferenceEvents.CONFERENCE_VISITOR_CODECS_CHANGED,
+            (codecList: CodecMimeType[]) => this._codecController.updateVisitorCodecs(codecList));
+        this._conference.on(
```

Comment:
Should these 2 be dbounced, to avoid a "storm" when joining last or when many people leave at once?

Reply:
I am still working on this since I need to test this with a larger call to figure out what the optimum wait time. We do not want it to be long since users might have to wait a long time for videos to start rendering if a codec switch on the sender is needed.